nilearn/_utils/niimg.py

```diff
@@ -36,15 +36,17 @@ def _safe_get_data(img, ensure_finite=False):
         If True, non-finite values such as (NaNs and infs) found in the
         image will be replaced by zeros.
 
+    copy_data: bool, default is False
+        If true, the returned data is a copy of the img data.
+
     Returns
     -------
     data: numpy array
         nilearn.image.get_data return from Nifti image.
     """
-    if hasattr(img, '_data_cache') and img._data_cache is None:
-        # By loading directly dataobj, we prevent caching if the data is
-        # memmaped. Preventing this side-effect can save memory in some cases.
+    if copy_data:
         img = copy.deepcopy(img)
+
```

Comment:
if the image did not have a `_data_cache`, the returned array is the (freshly filled in the call to `_get_data`) cache of the image itself, rather than a copy?

Reply:
@jeromedockes Yes, this was the behavior AFAICT. 

Here is an example script:

```python
import sys
import numpy as np
from nilearn import datasets
from nilearn.image import get_data
from nilearn._utils.niimg import _safe_get_data
from nilearn._utils.niimg_conversions import check_niimg

atlas_data = datasets.fetch_atlas_msdl() # Get some data
data = check_niimg(atlas_data.maps)
print("Ref count on data._data_cache = {0}".format(sys.getrefcount(data._data_cache)))
print("Initial sum = {0}".format(np.sum(data._data_cache)))
data2 = _safe_get_data(data)
data2[0,0,0,0] += 1
print("Ref count on data._data_cache = {0}".format(sys.getrefcount(data._data_cache)))
print("Sum for data._data_cache = {0}".format(np.sum(data._data_cache)))
print("Ref count on data2 = {0}".format(sys.getrefcount(data2)))
print("Sum for data2 = {0}".format(np.sum(data2)))
data._data_cache = None
data3 = _safe_get_data(data)
data3[0,0,0,0] += 2
print("Ref count on data3 = {0}".format(sys.getrefcount(data3)))
print("Sum for data3 = {0}".format(np.sum(data3)))
print("Sum for data2= {0}".format(np.sum(data2)))
```
Output:
```
Ref count on data._data_cache = 2
Initial sum = 3566.948371222825
Ref count on data._data_cache = 3
Sum for data._data_cache = 3567.948371222825
Ref count on data2 = 3
Sum for data2 = 3567.948371222825
Ref count on data3 = 2
Sum for data3 = 3568.948371222825
Sum for data2= 3567.948371222825
```